### PR TITLE
Add a macrobenchmark of bloom filter lookup

### PR DESCRIPTION
**Pull request by dcoutts**
With a setup resembling the LSM siutaion of many filters, in which we look up the same key.

Also add an initial version of a bulk lookup for the same key in many filters. This initial version just shares the hashing, but optimised ones using memory prefetching are also possible.

### CODE REVIEW

File: bloomfilter/Data/BloomFilter.hs [hunk 1]
```diff
@@ -180,6 +183,15 @@ elemCheapHashes ch ub = go 0 where
                                then go (i + 1)
                                else False
 
+-- | Query several Bloom filters for membership of a single key. The result
+-- is equivalent to @map (elem k) bs@ but may be faster when used with many
+-- filters by taking advantage of shared calculations.
+--
+elemMany :: Hashable a => a -> Vec.Vector (Bloom a) -> UVec.Vector Bool
```
Comment (phadej) [inline bloomfilter/Data/BloomFilter.hs:190]: doens't belong in the library. Why `Vec`, why not some other container? Oddly specific and ad-hoc.

File: src/Database/LSMTree/Internal/Run/BloomFilter.hs [hunk 2]
```diff
@@ -39,6 +39,7 @@ module Database.LSMTree.Internal.Run.BloomFilter (
     -- * Bloom
   , Bloom
   , Bloom.elem
+  , Bloom.elemMany
```
Comment (phadej) [inline src/Database/LSMTree/Internal/Run/BloomFilter.hs:42]: `elem` many should be rather defined in this (an lsm-tree specific module).

Comment (dcoutts) [reply src/Database/LSMTree/Internal/Run/BloomFilter.hs:42]: As noted above, the fast versions need access to the internal representation.

Comment (phadej) [reply src/Database/LSMTree/Internal/Run/BloomFilter.hs:42]: .. which is exported. E.g. the serialization uses it (and doesn't live in `bloomfilter`).

File: bench/macro/Bench/Database/LSMTree/Internal/BloomFilter.hs [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+      putStrLn "Subtract this from all subsequent benchmarks.)"
+      before <- getCurrentTime
+      evaluate (benchBaseline vbs rng0 benchmarkNumLookups)
+      after <- getCurrentTime
+      putStr "Finished: "
+      print (after `diffUTCTime` before)
+      putStrLn ""
+
+   do putStrLn "Benchmarking makeCheapHashes cost ... "
+      putStrLn "(This includes the cost of hashing the keys. Subtract this"
```
Comment (phadej) [inline bench/macro/Bench/Database/LSMTree/Internal/BloomFilter.hs:75]: Can't you do it for me?

**Review by jorisdral — APPROVED** [review]
LGTM

**Comment by dcoutts** [comment]
@phadej I've removed the `elemMany`, so it's just the benchmark code.

File: bench/macro/Bench/Database/LSMTree/Internal/BloomFilter.hs [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+-- | This gives us a combined cost of calculating the series of keys and their
+-- hashes.
+benchMakeCheapHashes :: Vector (Bloom SerialisedKey) -> StdGen -> Int -> ()
+benchMakeCheapHashes !_  !_   0 = ()
+benchMakeCheapHashes !bs !rng !n =
+    let k :: Word256
+        (!k, !rng') = uniform rng
+        !kh = Bloom.makeCheapHashes (serialiseKey k)
+     in kh `seq` benchMakeCheapHashes bs rng' (n-1)
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply bench/macro/Bench/Database/LSMTree/Internal/BloomFilter.hs:188]: Yes exactly. And to account for it we have a baseline benchmark that does _just_ the key generation, which we can use to subtract from the others.

File: bench/macro/Bench/Database/LSMTree/Internal/BloomFilter.hs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+lsmStyleBloomFilters l1 requestedFPR =
+    [ (numEntries, sizeFactor, numBits, numHashFuncs)
+    | (numEntries, sizeFactor)
+        <- replicate 4 (2^(l1+0), 1)   -- 4 runs at level 1 (tiering)
+        ++ replicate 8 (2^(l1+2), 4)   -- 8 runs at level 2 (tiering)
+        ++ replicate 8 (2^(l1+4),16)   -- 8 runs at level 3 (tiering)
+        ++            [(2^(l1+8),256)] -- 1 run  at level 4 (leveling)
+    , let numBits      = Bloom.monkeyBits numEntries requestedFPR
+          numHashFuncs = Bloom.monkeyHashFuncs numBits numEntries
+    ]
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply bench/macro/Bench/Database/LSMTree/Internal/BloomFilter.hs:112]: > Due to the incremental merges, most internal levels have 2x the number of runs (the incoming and the "normal" runs).

True, but I'm referring to the `sizeFactor` part of the quadruplet, where the `sizeFactor` is `256` even though in Monkey terminology the size factor would be 4 everywhere

> I've just realised though that it should also be 8 on the first level, sine although we count incoming runs as being on the same level, they're really the size of the smaller level above, so we should count them with the previous level.

:+1:

**Comment by dcoutts** [comment]
I've taken the suggestion from @phadej to subtract the baseline numbers in the presentation of results, and the suggestion from @jorisdral to also include allocations.

Current results on my box:
```
total number of entries:	 27787264
total filter size in bytes:	 28281750
total number of key lookups:	 25000000
Generating bloom filters...

Benchmarking baseline ... 
(This is the cost of just computing the keys.)
Finished.
Time net total:    3.29 seconds
Alloc net total:   1.18e11 bytes
Time net per key:  1.32e-7 seconds
Alloc net per key: 4704.00 bytes

Benchmarking makeCheapHashes ... 
(This includes the cost of hashing the keys, less the cost of computing the keys)
Finished.
Time net total:    0.71 seconds
Alloc net total:   4.00e8 bytes
Time net per key:  2.83e-8 seconds
Alloc net per key: 16.00 bytes

Benchmarking elemCheapHashes ... 
(this is the simple one-by-one lookup, less the cost of computing and hashing the keys)
Finished.
Time net total:    58.68 seconds
Alloc net total:   4.00e8 bytes
Time net per key:  2.35e-6 seconds
Alloc net per key: 16.00 bytes
```

**Review by jorisdral — APPROVED** [review]
LGTM

**Comment by jorisdral** [comment]
> I've taken the suggestion from @phadej to subtract the baseline numbers in the presentation of results, and the suggestion from @jorisdral to also include allocations.
> 
> Current results on my box:
> ...

For reference, this is what I'm getting:

<details>
<summary>Bench results</summary>

```
total number of entries:	 27787264
total filter size in bytes:	 28281750
total number of key lookups:	 25000000

Benchmarking baseline ...
(This is the cost of just computing the keys.)
Finished.
Time net total:    2.06 seconds
Alloc net total:   1.18e11 bytes
Time net per key:  8.24e-8 seconds
Alloc net per key: 4704.00 bytes

Benchmarking makeCheapHashes ...
(This includes the cost of hashing the keys, less the cost of computing the keys)
Finished.
Time net total:    0.65 seconds
Alloc net total:   4.00e8 bytes
Time net per key:  2.60e-8 seconds
Alloc net per key: 16.00 bytes

Benchmarking elemCheapHashes ...
(this is the simple one-by-one lookup, less the cost of computing and hashing the keys)
Finished.
Time net total:    24.74 seconds
Alloc net total:   4.00e8 bytes
Time net per key:  9.90e-7 seconds
Alloc net per key: 16.00 bytes
```

</details>

**Comment by dcoutts** [comment]
You've got a faster machine than me :-)

**Pull request merged**